Migrate initBlockWidget & inline script usage to Stimulus w-block controller

kituuu opened the pull request:
Related to #11597 
This PR introduced a new Stimulus controller, BlockController, with the identifier w-block, which would allow the unpacking and rendering of the unpacked block widget from Telepath JSON.

- Created BlockController.
- Removed old init method
- Written a new jest test for block controller.
- a few changes in existing python tests.
- manual testing on bakery site.


- streamfield not visible or working when the BlockController code is commented
<img width="1728" alt="Screenshot 2024-03-30 at 6 31 00 AM" src="https://github.com/wagtail/wagtail/assets/66073214/f7ae8419-bd09-442b-862d-0d5f4b93346f">

- streamfield working fine when BlockController is present.
<img width="1728" alt="Screenshot 2024-03-30 at 6 36 17 AM" src="https://github.com/wagtail/wagtail/assets/66073214/656ee60b-d2b2-481c-98fc-8c7bbbe5d71d">

kituuu commented:
@lb-

lb- commented on client/src/controllers/BlockController.ts:
```diff
@@ -0,0 +1,67 @@
+ * data-w-block-error-value="{error_json}">
+ * </div>
+ */
+
```
```suggestion
```

There should not be a line break after JSDoc blocks.

lb- commented on client/src/controllers/BlockController.ts:
```diff
@@ -0,0 +1,67 @@
+ */
+
+declare global {
+  interface Window {
```
Move this interface to the top of the file, just after the imports.

lb- commented on client/src/controllers/index.ts:
```diff
@@ -26,6 +26,7 @@ import { TeleportController } from './TeleportController';
 import { TooltipController } from './TooltipController';
 import { UnsavedController } from './UnsavedController';
 import { UpgradeController } from './UpgradeController';
+import { BlockController } from './BlockController';
```
Please add controller imports alphabetically, as per the comments in the file.

lb- commented on client/src/controllers/index.ts:
```diff
@@ -59,4 +60,5 @@ export const coreControllerDefinitions: Definition[] = [
   { controllerConstructor: TooltipController, identifier: 'w-tooltip' },
   { controllerConstructor: UnsavedController, identifier: 'w-unsaved' },
   { controllerConstructor: UpgradeController, identifier: 'w-upgrade' },
+  { controllerConstructor: BlockController, identifier: 'w-block' },
```
Please add the controller registration lines alphabetically, as per the comments at the start of this section.

lb- commented on client/src/controllers/BlockController.ts:
```diff
@@ -0,0 +1,67 @@
+
+export class BlockController extends Controller<HTMLElement> {
+  static values = {
+    initial: { type: object, default: {} }, // Initial value for the block
```
These comments can be set as JSDoc comments on the TypeScript declaration.

lb- commented on client/src/controllers/BlockController.ts:
```diff
@@ -0,0 +1,67 @@
+    error: object, // Block error (optional)
+  };
+
+  declare dataValue: object;
```
```suggestion
  /** Packed Telepath data. */
  declare dataValue: object;
```

As per comment above, move the description of values to JSDoc format and above the declaration lines.

lb- commented on client/src/controllers/BlockController.ts:
```diff
@@ -0,0 +1,67 @@
+  static values = {
+    initial: { type: object, default: {} }, // Initial value for the block
+    data: object, // Unpacked Telepath data
+    error: object, // Block error (optional)
```
Please give data and error values default like you have for initial, it's not critical but helps the code be more consistent and readable.

lb- commented on client/src/controllers/BlockController.ts:
```diff
@@ -0,0 +1,67 @@
+  }
+
+  static afterLoad() {
+    window.initBlockWidget = (id: string) => {
```
Just a reminder to remove the code from 

https://github.com/wagtail/wagtail/blob/f32c321cdd19f047d30449e493b22eff6adfe9b3/client/src/entrypoints/admin/telepath/blocks.js#L41-L63

lb- replied:
```suggestion
    /**
    * @deprecated RemovedInWagtail70
    */
    window.initBlockWidget = (id: string) => {
```

Let's also flag this as deprecated, it is only being added here to support backwards compatible usage of this global function.

lb- commented on client/src/controllers/BlockController.ts:
```diff
@@ -0,0 +1,67 @@
+  declare initialValue: object;
+
+  connect() {
+    const output = window.telepath.unpack(this.dataValue);
```
```suggestion
    const telepath = window.telepath;
    if ( !telepath) {
      // throw an error saying Telepath should exist. This way the code below can run safely
    }
    const output = window.telepath.unpack(this.dataValue);
```

lb- commented on client/src/controllers/BlockController.ts:
```diff
@@ -0,0 +1,67 @@
+
+  connect() {
+    const output = window.telepath.unpack(this.dataValue);
+    output.render(
```
```suggestion
    const element= this.element;
    if (!element.id) {
      // Throw an error here saying that the element needs an id. I think... Only of the block render code needs an id. 
    }
    const id = element.id;
    output.render(
```

Pull out the element to a variable earlier.

lb- reviewed (changes requested):
Awesome stuff @kituuu - for your first time doing a Stimulus PR & TypeScript - it's a great start. 

I've got a bunch of random feedback in this review, mostly small clean up items. 

I'd encourage you to read the following documentation sites to help you understand a bit more. 

1. Origin and core concepts of Stimulus  - https://stimulus.hotwired.dev/handbook/origin
2. Overview of a Stimulus controller https://stimulus.hotwired.dev/reference/controllers
3. Wagtail contributors guide for Stimulus https://docs.wagtail.org/en/stable/contributing/ui_guidelines.html#stimulus
4. Basic overview of Telepath API https://wagtail.github.io/telepath/api_reference/
5. Introduction to Jest (JS unit tests) - https://www.testim.io/blog/jest-testing-a-helpful-introductory-tutorial/

Good luck with the next steps.

lb- commented on client/src/controllers/BlockController.ts:
```diff
@@ -0,0 +1,67 @@
+import { object } from 'prop-types';
+
+/**
+ * Adds the ability to control a block element.
```
```suggestion
 * Adds the ability to unpack a Telepath object and render it on the controlled element.
```

lb- commented:
Telepath unfortunately does not have any TypeScript support but what you've done is good enough for now. Just get TypeScript happy and I'll see if I can look at a refined way to do this and help out.

kituuu commented:
@lb- I read the provided docs and stuff. I have some doubts, and I modified some things in the code. 
1. I tried writing some tests, which are working fine. It still needs to be well-optimised; any suggestions would be excellent. I will update the description later. For now can you suggest some more tests which needs to be implemented.
2. Earlier, we used `body.dataset.data` and `body.dataset.error`, but I observed that these are looking for `data-data` and `data-error`. As suggested in the code snippet, we were using  `data-w-block-data-value`, `data-w-block-initial-value` so the correct selector should be `body.dataset.wBlockDataValue` (something I observed while writing jest tests), so I updated code.
3. Also, some tests are failing for eg this one 
```
======================================================================
FAIL: test_create_alias_without_publish_permissions (wagtail.admin.tests.api.test_pages.TestCreatePageAliasAction.test_create_alias_without_publish_permissions)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/opt/hostedtoolcache/Python/3.12.2/x64/lib/python3.12/unittest/case.py", line 58, in testPartExecutor
    yield
  File "/opt/hostedtoolcache/Python/3.12.2/x64/lib/python3.12/unittest/case.py", line 634, in run
    self._callTestMethod(testMethod)
    ^^^^^^^^^^^^^^^^^
  File "/opt/hostedtoolcache/Python/3.12.2/x64/lib/python3.12/unittest/case.py", line 589, in _callTestMethod
    if method() is not None:
    ^^^^^^^^^^^^^^^^^
  File "/home/runner/work/wagtail/wagtail/wagtail/admin/tests/api/test_pages.py", line 1957, in test_create_alias_without_publish_permissions
    self.assertEqual(
  File "/opt/hostedtoolcache/Python/3.12.2/x64/lib/python3.12/unittest/case.py", line 885, in assertEqual
    assertion_func(first, second, msg=msg)
    ^^^^^^^^^^^^^^^^^
  File "/opt/hostedtoolcache/Python/3.12.2/x64/lib/python3.12/unittest/case.py", line 1184, in assertDictEqual
    self.fail(self._formatMessage(msg, standardMsg))
    ^^^^^^^^^^^^^^^^^
  File "/opt/hostedtoolcache/Python/3.12.2/x64/lib/python3.12/unittest/case.py", line 715, in fail
    raise self.failureException(msg)
    ^^^^^^^^^^^^^^^^^
AssertionError: {'detail': 'You do not have permission to publish a page at the destination'} != {'detail': 'You do not have permission to perform this action.'}
- {'detail': 'You do not have permission to publish a page at the destination'}
?                                            ^^^  -  ^^^^^^^^^^^^^^^^^^^^

+ {'detail': 'You do not have permission to perform this action.'}
?                                            ^^^^^^^^^    ^    +
```
but when I ran this same test on my system it is passing.

```
(base) ➜  wagtail git:(kitu/initBlockWidgetMigration) python runtests.py wagtail.admin.tests.api.test_pages.TestCreatePageAliasAction.test_create_alias_without_publish_permissions
Found 1 test(s).
Creating test database for alias 'default'...
/Users/kitu/Developer/opensource/wgtail/wagtail/wagtail/admin/rich_text/editors/draftail/__init__.py:55: RuntimeWarning: Draftail received an unknown feature 'made-up-feature'.
  warnings.warn(
System check identified some issues:

WARNINGS:
tests.TaggedChildPage: (wagtailsearch.W001) Core Page fields missing in `search_fields`
        HINT: Ensure that TaggedChildPage extends the Page model search fields `search_fields = Page.search_fields + [...]`
tests.TaggedGrandchildPage: (wagtailsearch.W001) Core Page fields missing in `search_fields`
        HINT: Ensure that TaggedGrandchildPage extends the Page model search fields `search_fields = Page.search_fields + [...]`
tests.TaggedPage: (wagtailsearch.W001) Core Page fields missing in `search_fields`
        HINT: Ensure that TaggedPage extends the Page model search fields `search_fields = Page.search_fields + [...]`

System check identified 3 issues (0 silenced).
.
----------------------------------------------------------------------
Ran 1 test in 0.073s

OK
Destroying test database for alias 'default'...

```
Can you help me with these failed ci/cd tests?

lb- commented on client/src/controllers/BlockController.ts:
```diff
@@ -0,0 +1,83 @@
+    this.dispatch('ready', { detail: { ...output }, cancelable: false });
+  }
+
+  static afterLoad() {
```
@kituuu  I think you may misunderstand the purpose of the code in `afterLoad`, that code will run when the Stimulus controller will be registered.

This is not meant to use the NEW data attributes, but rather gives a way for the OLD (being removed) approach to be used by custom code. We are setting this up so that we can remove the equivalent code that existed in `client/src/entrypoints/admin/telepath/blocks.js` and it should 100% be exactly the same.

The reason we are 'moving' it to this new code is so that we can track it being removed fully once we bump to the next major release.

It also stops other code from trying to use this window global function which will be removed (either our own admin code or otherwise).

As such, we probably do not even need a unit test for the `afterLoad` code, it's not tested today and we will remove it in < a year so no need to add tests.

kituuu replied:
Oh okay, so should I just revert the changes in `afterLoad` and remove the `BlockController.test.js`? @lb-

lb- replied:
yep. I have a suggested test coming with another review.

lb- commented on client/src/controllers/BlockController.ts:
```diff
@@ -0,0 +1,83 @@
+  }
+}
+
+/**
+ * Adds the ability to unpack a Telepath object and render it on the controlled element.
+ *
+ * @example
+ * <div
+ * id="{id}"
+ * data-block
+ * data-controller="w-block"
+ * data-w-block-data-value="{block_json}"
+ * data-w-block-initial-value="{value_json}"
+ * data-w-block-error-value="{error_json}">
+ * </div>
+ */
+
+export class BlockController extends Controller<HTMLElement> {
```
There is still a line break after the JSDoc, here's a suggested fix with some more refinements to the example HTML also. We do not need much in this example but need to assume that the Django template syntax is not to be used in the example.

```suggestion
/**
 * Adds the ability to unpack a Telepath object and render it on the controlled element.
 *
 * @example
 * <div
 *  id="some-id"
 *  data-controller="w-block"
 *  data-w-block-data-value='{"some":"json"}'
 * >
 * </div>
 */
export class BlockController extends Controller<HTMLElement> {
```

lb- commented on client/src/controllers/BlockController.ts:
```diff
@@ -0,0 +1,83 @@
+ */
+
+export class BlockController extends Controller<HTMLElement> {
+  /**
+   * initial - Initial value for the block
+   * data - Unpacked Telepath data
+   * error - Block error (optional)
+   */
+  static values = {
+    initial: { type: object, default: undefined },
+    data: { type: object, default: undefined },
+    error: { type: object, default: undefined },
+  };
```
OK, so the current approach as is will not work correctly, we should be using `Object` not `object` as this is the Stimulus value declaration. 

Also, it appears that this code does not work as is, so you will need to use `initial` as an array it seems, but please double check.

Have a good read of https://stimulus.hotwired.dev/reference/values to understand what's going on here.

```suggestion
  static values = {
    initial: { type: Array, default: [] },
    error: { type: Object, default: {} },
    data: { type: Object, default: {} },
  };
```

Also the JSDoc should not be on the static property but in the TypeScript declarations below.

lb- commented on client/src/controllers/BlockController.ts:
```diff
@@ -0,0 +1,83 @@
+    error: { type: object, default: undefined },
+  };
+
+  /** Packed Telepath data. */
+  declare dataValue: object;
+  declare errorValue: object;
+  declare initialValue: object;
```
Just a clean up, we still need lowercase `object` here but I suggest we re-order this to alpha and add a basic JSDoc to each.

```suggestion
  /** Initial value. */
  declare initialValue: object;
  /** Optional error value. */
  declare errorValue: object;
  /** Block data value to unpack with Telepath. */
  declare dataValue: object;
```

lb- commented on client/src/controllers/BlockController.ts:
```diff
@@ -0,0 +1,83 @@
+  declare errorValue: object;
+  declare initialValue: object;
+
+  connect() {
```
Here's a suggested `connect` method, code is good but just a bit of a refinement.

1. Check for the id first, then throw if not present.
2. Check for Telepath second, then throw if not present.
3. Avoid getting the `element.id` twice, just set as a variable and use throughout the code.


```ts

  connect() {
    const element = this.element;
    const id = element.id;

    if (!id) throw new Error('Block element needs an id');

    const telepath = window.telepath;

    if (!telepath) throw new Error("Telepath doesn't exit");

    const output = telepath.unpack(this.dataValue);
    output.render(element, id, this.initialValue, this.errorValue);

    this.dispatch('ready', { detail: { ...output }, cancelable: false });
  }
```

lb- commented on client/src/controllers/BlockController.test.js:
```diff
@@ -0,0 +1,76 @@
+import { Application } from '@hotwired/stimulus';
+import { BlockController } from './BlockController'; // Replace with your controller path
```
Here's an updated starting point for the tests, well done on getting this mostly working, hopefully this helps though.

```js
import { Application } from '@hotwired/stimulus';
import { BlockController } from './BlockController';

const render = jest.fn();
const unpack = jest.fn(() => ({ render }));
window.telepath = { unpack };

describe('BlockController', () => {
  const eventNames = ['w-block:ready'];

  const events = {};

  eventNames.forEach((name) => {
    document.addEventListener(name, (event) => {
      events[name].push(event);
    });
  });

  let application;
  let errors = [];

  const setup = (html, { identifier = 'w-block' } = {}) => {
    document.body.innerHTML = `<main>${html}</main>`;

    application = new Application();

    application.register(identifier, BlockController);

    application.handleError = (error, message) => {
      errors.push({ error, message });
    };

    application.start();

    return Promise.resolve();
  };

  beforeEach(() => {
    application?.stop();
    document.body.innerHTML = '';
    errors = [];
    eventNames.forEach((name) => {
      events[name] = [];
    });
    jest.clearAllMocks();
  });

  it('does nothing if block element is not found', async () => {
    await setup('<div></div>');

    expect(errors).toHaveLength(0);

    expect(unpack).not.toHaveBeenCalled();

    expect(events['w-block:ready']).toHaveLength(0);
  });

  it('should render block if element is controlled', async () => {
    const data = { name: 'John Doe' };

    await setup(
      `<div id="my-element" data-controller="w-block" data-w-block-data-value='${JSON.stringify({ name: 'John Doe' })}'></div>`,
    );

    expect(errors).toHaveLength(0);
    expect(unpack).toHaveBeenCalledWith(data);
    expect(render).toHaveBeenCalledWith(
      document.getElementById('my-element'),
      'my-element',
      [],
      {},
    );
    expect(events['w-block:ready']).toHaveLength(1);
  });

  it('should call the unpacked render function with provided initial & error data', async () => {
    /** */
  });

  it('should throw an error if used on an element without an id', async () => {
    /** */
    // expect(errors).toHaveLength(1);
    // expect(errors).toHaveProperty('0.messages', '???');
  });

  it('should throw an error if Telepath is not available in the window global', async () => {
    /** */
    // expect(errors).toHaveLength(1);
    // expect(errors).toHaveProperty('0.messages', '???');
  });
});
```

lb- commented on wagtail/blocks/base.py:
```diff
@@ -568,10 +568,7 @@ def render_with_errors(self, name, value, attrs=None, errors=None, renderer=None
 
```
In the line above this, it looks like we will have to change the default error json to an empty object.

```py
            error_json = json.dumps({})
```

@gasman  - I can do some digging but do you know if `value_json` will usually be a List and `error_json` will usually be an ?? (List) or can they be any structure (inc `None`)?

lb- reviewed (commented):
Did another review, this code was not working for me so be sure to do some manual testing also to check different scenarios.

Hopefully this next round of comments helps.

kituuu commented:
Bit busy these days. Will push work in a day

lb- commented on client/src/controllers/BlockController.ts:
```diff
@@ -0,0 +1,76 @@
+  declare errorValue: Object;
+  /** Block data value to unpack with Telepath. */
+  // eslint-disable-next-line @typescript-eslint/ban-types
+  declare dataValue: Object;
```
The Typescript type should be `object` (lowercase). 

It's only the Stimulus value type that should be the upper case `Object`. 

The reason for this is that the Stimulus one needs a constructor class, but Typescript does not use that.

kituuu replied:
Oh, okie. I got it.

kituuu commented:
> Did another review, this code was not working for me so be sure to do some manual testing also to check different scenarios.
> 
> Hopefully this next round of comments helps.

@lb- Can you tell me how to do manual testing? Also, I need help finding out why this `backend ci/cd` failed tests. Because in my system, I did not get those failed tests when I ran the `"make test"`.

lb- commented:
@kituuu you can safely ignore the one failing test at the moment, it's a package change issue and has nothing to do with your code. 

![Screenshot_20240325-055156](https://github.com/wagtail/wagtail/assets/1396140/eb257639-9e3a-4e83-9ea0-954bc5fd1b46)

As for manual testing, anything that uses a StreamField in your local Bakerydemo setup is fine. Try new values (E.g. creating a new page) and editing. Check that the StreamField works correctly. 

Once you feel confident it's ready for another review, just comment and I'll take a look also. 

Great work so far here. It's a bit of a learning curve I know.

lb- commented on client/src/controllers/BlockController.ts:
```diff
@@ -0,0 +1,74 @@
+    initial: { type: Array, default: [] },
+    error: { type: Object, default: {} },
+    data: { type: Object, default: {} },
+  };
```
Been thinking about this a bit and I feel like we might get a better, simpler, solution if we change this to have TWO values only.

```
  static values = {
    args: { type: Array, default: [] },
    data: { type: Object, default: {} },
  };
```

The `initial`, and `error` could be provided as a single Stimulus value, in an array format.

kituuu replied:
okay

lb- commented on client/src/controllers/BlockController.ts:
```diff
@@ -0,0 +1,74 @@
+    if (!telepath) throw new Error("Telepath doesn't exit");
+
+    const output = telepath.unpack(this.dataValue);
+    output.render(element, id, this.initialValue, this.errorValue);
```
We would then replace the call here with a JS Spread of the `argumentsValue`.

```suggestion
    output.render(element, id, ...this.argumentsValue);
```

lb- commented on wagtail/blocks/base.py:
```diff
@@ -564,14 +564,11 @@ def render_with_errors(self, name, value, attrs=None, errors=None, renderer=None
             error_json = json.dumps(get_error_json_data(error))
         else:
-            error_json = "null"
+            error_json = json.dumps({})
```
```suggestion
            error_json = "null"
```

This could change back to null or 

```suggestion
            error_json = json.dumps(None)
```

lb- commented on wagtail/blocks/base.py:
```diff
@@ -564,14 +564,11 @@ def render_with_errors(self, name, value, attrs=None, errors=None, renderer=None
-                <script>
-                    initBlockWidget('{id}');
-                </script>
+                <div id="{id}" data-block data-controller="w-block" data-w-block-data-value="{block_json}" data-w-block-initial-value="{value_json}" data-w-block-error-value="{error_json}"></div>
```
```suggestion
                <div id="{id}" data-block data-controller="w-block" data-w-block-data-value="{block_json}" data-w-block-arguments-value="[{value_json},{error_json}]"></div>
```
Finally, we would provide a single `arguments` value that is an array `[...,...]` value with the two JSON strings.

lb- reviewed (commented):
Hey @kituuu  sorry to add to the noise here but I actually think we should take a slightly different approach to handling of the `error` and the `initial` value.

Instead of passing these in as separate items, passing these in as one array `arguments` gives a bit more flexibility and it also will ensure that if the error/initial values are not objects we do not need to worry about the way Stimulus parses these.

I have added three other comments with this review block, hopefully they make sense. You will need to make other changes to the unit tests and TypeScript types to suit though.

kituuu commented:
Hey, @lb- Ig, you can review it. I have manually checked the work. First time by commenting on the connect and afterload functions. The second time, by keeping these functions. The stream field blocks only worked in the second case, so it works fine ...

lb- commented:
Thanks @kituuu I'll try to do a full review next week. I'll see if there are any quick comments I can do now though.

lb- commented on client/src/controllers/BlockController.ts:
```diff
@@ -0,0 +1,71 @@
+
+declare global {
+  interface Window {
+    initBlockWidget?: (id: string) => void; // Declare the initBlockWidget function as optional on the Window interface
```
```suggestion
    initBlockWidget?: (id: string) => void;
```

No need for this extra comment.

lb- commented on wagtail/admin/tests/pages/test_create_page.py:
```diff
@@ -1494,7 +1494,7 @@ def test_inline_streamfield(self):
 
         # response should include HTML declarations for streamfield child blocks
-        self.assertContains(response, '<div id="sections-__prefix__-body" data-block="')
+        self.assertContains(response, '<div id="sections-__prefix__-body" data-block')
```
```suggestion
        self.assertContains(response, '<div id="sections-__prefix__-body" data-block')
        soup = this.get_soup(response.conent)
        blockDiv = soup.find('div', {"data-controller":"w-block"})
        self.assertIsNotNone(blockDiv)
        # also check the div has an attribute for the two data values, just that they exist, not the content inside 
        
```

It would be good to bolster this test a bit more, leave the code better than we found it. 

We should test the data attributes are being used for the Stimulus controller. No need to test the JSON values, just that the attributes are used. 

Search for other tests in the codebase that use soup, see the docs here https://beautiful-soup-4.readthedocs.io/en/latest/

kituuu replied:
```python
self.assertContains(response, '<div id="sections-__prefix__-body" data-block')
soup = self.get_soup(response.content)
blockDiv = soup.find('div', {"data-controller":"w-block"})
self.assertIsNotNone(blockDiv)
# block div should contain this attributes
self.assertTrue(blockDiv.has_attr('data-w-block-arguments-value'))
self.assertTrue(blockDiv.has_attr('data-w-block-data-value'))
```
Something like this will do the job. What do you think?

lb- replied:
Love it

kituuu replied:
Hey, @lb- Can you give me few more issues to work on...

lb- replied:
See if you're up for tackling https://github.com/wagtail/wagtail/issues/11537 (or even part of it). If you have an idea on the approach, comment on the issue and go for it. 

There may be others but that's all I can think of now that might be a good progression from this issue. 

Another thing to do is go through unconfirmed bugs and see if any need to be reproduced, helps the community with triage but also you may find something you can give a go at fixing. Thanks for being keen on staying involved.

lb- reviewed (commented):
Ok. Codes looking really solid, a few little suggestions and then I think you're good to make that a non-draft. 

Please also update the PR description with a summary of the change, a screenshot of a block being used (maybe with the Dev tools open showing the new data attributes in the response HTML). 

I'll do another cross check next week with the plan to merge, but we'll give any other core ram members a chance to double check the approach. 

Incredible work here, thank you!

lb- reviewed (approved):
Thank you @kituuu  - incredible stuff here!

This is working great for me, tested a bunch of different scenarios.

I have made a few minor tweaks and will push these up (add release notes, update the JSDoc examples & test example data to be closer to real-world usage, update unit tests to check the error message, a few refinements to the JSdoc text generally).

Thank you so much for pushing through on this and helping out with the path to full CSP support in Wagtail!
